examples_tests/48.ArithmeticUnitTest/main.cpp

```diff
@@ -0,0 +1,230 @@
+#define _IRR_STATIC_LIB_
+#include <irrlicht.h>
+
+#include "../../source/Irrlicht/COpenGLDriver.h"
+
+
+using namespace irr;
+using namespace core;
+using namespace video;
+using namespace asset;
+
+enum TestOperation {
+	TO_AND = 1,
+	TO_XOR,
+	TO_OR,
+	TO_ADD,
+	TO_MUL,
+	TO_MIN,
+	TO_MAX
+};
+
+template<typename T, typename OP, uint32_t subgroupSize>
+T emulatedSubgroupReduction(const T* data, const uint32_t globalInvocationIndex)
+{
+	auto subgroupID = globalInvocationIndex / subgroupSize;
+	//auto subgroupOffset = data + subgroupID * subgroupSize;
+	T retval = data[0];
+	for (auto i = 1u; i < subgroupSize; i++)
+		retval = OP(retval, data[i]);
+	return retval;
+}
+uint32_t add(uint32_t a, uint32_t b)
+{
+	return a + b;
+}
+
+constexpr uint32_t BUFFER_SIZE = 128u * 1024u * 1024u;
+//returns true if result matches
+template<typename Func>
+bool validateResults(video::IVideoDriver* driver, const uint32_t* inputData, video::IGPUBuffer* bufferToDownload)
+{
+	constexpr uint64_t timeoutInNanoSeconds = 15000000000u;
+	const uint32_t alignment = sizeof(uint32_t);
+	auto downloadStagingArea = driver->getDefaultDownStreamingBuffer();
+	auto downBuffer = downloadStagingArea->getBuffer();
+
+
+
+	uint32_t address = std::remove_pointer<decltype(downloadStagingArea)>::type::invalid_address;
+	auto unallocatedSize = downloadStagingArea->multi_alloc(1u, &address, &BUFFER_SIZE, &alignment);
+	if (unallocatedSize)
+	{
+		os::Printer::log("Could not download the buffer from the GPU!", ELL_ERROR);
+		return false;
+	}
+	driver->copyBuffer(bufferToDownload, downBuffer, 0, address, BUFFER_SIZE);
+	auto downloadFence = driver->placeFence(true);
+	auto result = downloadFence->waitCPU(timeoutInNanoSeconds, true);
+	if (result == video::E_DRIVER_FENCE_RETVAL::EDFR_TIMEOUT_EXPIRED || result == video::E_DRIVER_FENCE_RETVAL::EDFR_FAIL)
+	{
+		os::Printer::log("Could not download the buffer from the GPU, fence not signalled!", ELL_ERROR);
+		downloadStagingArea->multi_free(1u, &address, &BUFFER_SIZE, nullptr);
+		return false;
+	}
+	if (downloadStagingArea->needsManualFlushOrInvalidate())
+		driver->invalidateMappedMemoryRanges({ {downloadStagingArea->getBuffer()->getBoundMemory(),address,BUFFER_SIZE} });
+
+	auto dataFromBuffer = reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(downloadStagingArea->getBufferPointer()) + address);
+	//now check if the data obtained has valid values
+	uint32_t* end = dataFromBuffer + BUFFER_SIZE / sizeof(uint32_t);
+	uint32_t invocationIndex = 0;
+	for (uint32_t* i = dataFromBuffer; i < end; i++)
+	{
+		uint32_t val = *(inputData + invocationIndex);	//TODO instead the template function and compare result
```

Comment:
Here i need to emulate what is being done on the GPU and then compare it to the downloaded result